src/transactions/ParallelApplyUtils.h

```diff
@@ -80,20 +75,189 @@ void preParallelApplyAndCollectModifiedClassicEntries(
+
+void recordModifiedAndRestoredEntries(
+    SearchableSnapshotConstPtr liveSnapshot, ParallelApplyEntryMap& entryMap,
+    UnorderedMap<LedgerKey, uint32_t>& roTTLBumps,
+    RestoredEntries& threadRestoredEntries, TxBundle const& txBundle,
+    ParallelTxReturnVal const& res);
+
 std::optional<LedgerEntry> getLiveEntry(LedgerKey const& lk,
                                         SearchableSnapshotConstPtr liveSnapshot,
                                         ParallelApplyEntryMap const& entryMap);
 
+class ThreadParallelApplyLedgerState;
+class GlobalParallelApplyLedgerState
+{
+
+    // Contains the hot archive state from the start of the ledger close. If a
+    // key is in here, it is "evicted". An invariant is that if a key is in here
+    // it is _not_ in the live snapshot.
+    SearchableHotArchiveSnapshotConstPtr mHotArchiveSnapshot;
+
+    // Contains the live soroban state from the start of the ledger close. If a
+    // key is in here, it is either "archived" or "live", depending on its TTL.
+    // Classic entries are always live, soroban entries always have an
+    // associated TTL entry and if the TTL is in the past the entry is
+    // "archived", otherwise "live". An invariant is that if a key is in here it
+    // is _not_ in the hot archive snapshot.
+    SearchableSnapshotConstPtr mLiveSnapshot;
+
+    // Contains restorations that happened during each stage of the parallel
+    // soroban phase. As with mGlobalEntryMap, this is propagated stage to stage
+    // by being split into per-thread maps and re-merged at the end of the
+    // stage, before begin committed to the ltx at the end of the phase. As with
+    // restorations inside the thread, these entries are the restored values _at
+    // their time of restoration_ which may be further overridden by
+    // mGlobalEntryMap.
+    RestoredEntries mGlobalRestoredEntries;
+
+    // Contains two different sets of entries:
+    //
+    //  - Classic entries modified during earlier sequential phases that are
+    //    read during the parallel soroban phase. These are copied out of the
+    //    ltx before the parallel soroban phase begins.
+    //
+    //  - Dirty entries resulting from each stage of the parallel soroban phase.
+    //    These are propagated from stage to stage of the parallel soroban phase
+    //    -- split into disjoint per-thread maps during execution and merged
+    //    after -- as well as written back to the ltx at the phase's end.
+    ParallelApplyEntryMap mGlobalEntryMap;
+
+  public:
+    GlobalParallelApplyLedgerState(AppConnector& app, AbstractLedgerTxn& ltx,
+                                   std::vector<ApplyStage> const& stages);
+
+    std::optional<LedgerEntry> getLiveEntryOpt(LedgerKey const& key) const;
+    void upsertEntry(LedgerKey const& key, LedgerEntry const& entry);
+    void eraseEntry(LedgerKey const& key);
+
+    RestoredEntries const& getRestoredEntries() const;
+
+    void commitChangesFromThread(AppConnector& app,
```

Comment:
nit: It seems to me like this should be private - it's suspicious that we would ever want to flush changes from a single thread.

Reply:
oh, yes, true.